application/common/models/EmailQueue.php

```diff
@@ -0,0 +1,219 @@
+        $eventLogDetails = null
+    ) {
+        $emailQueue = new EmailQueue();
+        $emailQueue->to_address = $toAddress;
+        $emailQueue->subject = $subject;
+        $emailQueue->text_body = $textBody;
+        $emailQueue->html_body = $htmlBody;
+        $emailQueue->cc_address = $ccAddress;
+        $emailQueue->event_log_user_id = $eventLogUserId;
+        $emailQueue->event_log_topic = $eventLogTopic;
+        $emailQueue->event_log_details = $eventLogDetails;
+
+        $emailQueue->send();
+
+        return $emailQueue;
+    }
+
+    /**
+     * Attempt to send email. Returns true on success or throws exception.
+     * @return void
+     * @throws \Exception
+     */
+    public function send()
+    {
+        $log = [
+            'class' => __CLASS__,
+            'action' => 'send',
+            'to' => $this->to_address,
+            'subject' => $this->subject,
+        ];
+
+        /*
+         * Try to send email or throw exception
+         */
+        try {
+            $mailer = $this->getMailer();
+            $sent = $mailer->send();
+            if ($sent === 0 || $sent === false) {
+                throw new \Exception('Unable to send email', 1461011826);
+            }
+
+            /*
+             * Create event log entry if needed
+             */
+            $this->createEventLogEntry();
+
+            /*
+             * Remove entry from queue (if saved to queue) after successful send
+             */
+            $this->removeFromQueue();
+
+            /*
+             * Log success
+             */
+            $log['status'] = 'sent';
+            \Yii::warning($log, 'application');
+
+        } catch (\Exception $e) {
+            /*
+             * Send failed, attempt to queue
+             */
+            $this->attempts_count += 1;
+            $this->last_attempt = Utils::getDatetime();
+            $this->queue($log, $e);
+        }
+    }
+
+    /**
+     * Builds a mailer object from $this and returns it
+     * @return \yii\mail\MessageInterface
+     */
+    private function getMailer()
+    {
+        $mailer = \Yii::$app->mailer->compose();
+        $mailer->setFrom(\Yii::$app->params['fromEmail']);
+        $mailer->setTo($this->to_address);
+        $mailer->setSubject($this->subject);
+        $mailer->setTextBody($this->text_body);
+
+        /*
+         * Conditionally set optional fields
+         */
+        $setMethods = [
+            'setCc' => $this->cc_address,
+            'setHtmlBody' => $this->html_body,
+        ];
+        foreach ($setMethods as $method => $value) {
+            if ($value) {
+                $mailer->$method($value);
+            }
+        }
+        
+        return $mailer;
+    }
+
+    /**
+     * Creates an EventLog entry if $this->event_log_* properties are set
+     */
+    private function createEventLogEntry()
+    {
+        if ($this->event_log_topic !== null && $this->event_log_details !== null && $this->event_log_user_id !== null) {
```

Comment:
Sure, I'll have it return false in an ```else```. If there is an error creating the entry it throws an exception.